src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1058,30 +1129,66 @@ private static bool CheckLocalVariable(CSharpSyntaxNode tree, LocalSymbol local,
+                ErrorCode.ERR_RefReadonlyLocalCause, 
                 ErrorCode.ERR_AssgReadonlyLocalCause,
+
+                ErrorCode.ERR_RefReadonlyLocal2Cause,
+                // impossible since readonly locals are never byref, but would be a reasonable error otherwise
                 ErrorCode.ERR_RefReadonlyLocal2Cause,
                 ErrorCode.ERR_AssgReadonlyLocal2Cause
             };
 
-            int index = (cause != 0 ? 4 : 0) + (checkingReceiver ? 2 : 0) + (kind == BindValueKind.OutParameter ? 0 : 1);
+            int index = (checkingReceiver ? 3 : 0) + (kind == BindValueKind.RefOrOut ? 0 : (kind == BindValueKind.RefReturn ? 1 : 2));
+
+            Error(diagnostics, ReadOnlyLocalErrors[index], tree, local, cause.Localize());
 
-            Debug.Assert(index != 2 && index != 3);
-            // There is no way that we can have no cause AND a read-only local nested in a struct with a
-            // writable field. What would make the local read-only if not one of the causes above?  (Const
-            // locals may not be structs, so we would already have errored out in that scenario.)
+            return false;
+        }
+
+        private bool CheckIsCallVariable(BoundCall call, CSharpSyntaxNode node, BindValueKind kind, bool checkingReceiver, DiagnosticBag diagnostics)
+        {
+            // A call can only be a variable if it returns by reference. If this is the case,
+            // whether or not it is a valid variable depends on whether or not the call is the
+            // RHS of a return or an assign by reference:
+            // - If call is used in a context demanding ref-returnable reference all of its ref
+            //   inputs must be ref-returnable
 
-            if (cause != 0)
+            var methodSymbol = call.Method;
+            if (methodSymbol.RefKind != RefKind.None)
             {
-                Error(diagnostics, ReadOnlyLocalErrors[index], tree, local, cause.Localize());
+                if (kind == BindValueKind.RefReturn)
+                {
+                    var args = call.Arguments;
+                    var argRefKinds = call.ArgumentRefKindsOpt;
+                    if (!argRefKinds.IsDefault)
+                    {
+                        for (var i = 0; i < args.Length; i++)
+                        {
+                            if (argRefKinds[i] != RefKind.None && !CheckIsVariable(args[i].Syntax, args[i], kind, false, diagnostics))
+                            {
+                                var errorCode = checkingReceiver ? ErrorCode.ERR_RefReturnCall2 : ErrorCode.ERR_RefReturnCall;
+                                var parameterIndex = call.ArgsToParamsOpt.IsDefault ? i : call.ArgsToParamsOpt[i];
+                                var parameterName = methodSymbol.Parameters[parameterIndex].Name;
+                                Error(diagnostics, errorCode, call.Syntax, methodSymbol, parameterName);
```

Comment:
Yes. I tried dropping errors from arguments and it becomes really hard to figure what went wrong and what action to take to fix the error.